### ISSUE TITLE
Verifier: Improve floating-point column result matching

### ISSUE THREAD

**Issue by caithagoras** [prestodb/presto#12229]
Currently, verifier rounds floating point to n digit (n=3 by default) and checksum the rounded value. This may cause false mismatch if the absolute values are large, and may clause false match if the absolute values are close to 0.

Instead, we might prefer to use relative errors of the column average (ref: http://floating-point-gui.de/errors/comparison/)

Doing so requires NaN and Infinity be handled explicitly.
https://github.com/prestodb/hql-translator/blob/master/validator/src/main/java/com/facebook/presto/hqltranslator/validator/matcher/checksum/FloatingPointColumnValidator.java#L95

**Comment by findepi** [comment]
I understand the reason for rounding is that result of `double` computation may depend on the order in which it's done.

> Instead, we might prefer to use relative errors of the column average (ref: http://floating-point-gui.de/errors/comparison/)

Consider the case when average is small and a particular value is large.
Since the accepted error is going to be relative to average, for this particular, large value, the accepted error will be tiny.

Instead, I would suggest rounding to a fixed number (eg 8) of most significant digits.

**Comment by caithagoras** [comment]
Another issue we're trying to improve here is the reporting of the mismatches, to provide release oncall with insights about what the issues might be. Reporting averages, Nan count, and infinity counts would help, but a checksum mismatch provide very minimal actionable items for the release oncall.
Thoughts?  @wenleix @mbasmanova

**Comment by mbasmanova** [comment]
I don't believe trying to improve the comparison of doubles is worth the effort. Even if we solve the issue with `double` columns, we still have issues with complex types that use doubles such as `map(string, double)`. Furthermore, I've seen a lot of queries where doubles appear as varchar because the query is doing `cast(d as varchar)`. Therefore, I think the most effective solution is to retry control query a few times to see if it also produces different results. Since retrying is time- and resource-consuming, I suggest to retry control query only after test query completed with different results. Thoughts?

**Comment by caithagoras** [comment]
@mbasmanova I don't see how this is related to the retry of control query or non-determinism. This is a separate improvement dealing with floating point columns for deterministic queries. Yes, the case with complex types or double casting to varchar is still not addressed by it, but our goal is to minimize manual effort in failure investigation in the verification process and therefore I think it worth the effort (1-2 commits).

I don't think there is only a single path of improvement. The retry for non-deterministic `INSERT` queries can and should be quickly addressed, but even so it will only mitigate a portion of the failures, and in order to mitigate the majority of low-signal failures, we need to do small improvements here and there.

**Comment by mbasmanova** [comment]
@caithagoras 

> Reporting averages, Nan count, and infinity counts would help, but a checksum mismatch provide very limited actionable items for the release oncall.

The number of queries with double columns that fail correctness check is too large. I don't think adding more information about how different the results are would help as this still requires manual inspection. Can you think of a way to automate the verification and provide HIGH signal result?

Alternatively, would you consider adding a configuration flag which tells the Verifier to skip comparisons for column types that include `double`?

**Comment by caithagoras** [comment]
> The number of queries with double columns that fail correctness check is too large.
That's before the fix for non-deterministic `INSERT` queries and floating point column handling. I would expect the # of floating point columns (double and float) mismatch dramatically reduced after those 2 fixes. At that stage, those information would be handful if only tiny amount of result mismatches are for double columns.

**Comment by caithagoras** [comment]
Addressed in https://github.com/prestodb/presto/commit/c4356e3a486c324c7416e842f07f046581b232a7
